src/drivers/fluid_winmidi.c

```diff
@@ -27,35 +27,85 @@
+ * the settings midi.winmidi.maxdevices and midi.winmidi.device.
+ * For example, let the following device names:
+ * 0:Port MIDI SB Live! [CE00], 1:SB PCI External MIDI, default, multi:x[ y z ..]
+ * The setting midi.winmidi.maxdevices must be set to the number of devices to
+ * work with. This allows the driver to check if the real devices number exist
+ * and allocate only the necessary memory.
+ * Then the driver is able receive MIDI messages comming from distinct devices
+ * and forward these messages on distinct MIDI channels set.
+ * 1.1)For example, if the user chooses 2 devices at index 0 and 1, the user
+ * must specify this by midi.winmidi.maxdevices set to 2 and putting the name
+ * "multi:0 1" in midi.winmidi.device setting.
+ * We get a fictif device composed of real devices (0,1). This fictif device
+ * behaves like a device with 32 MIDI channels whose messages are forwarded to
+ * driver output as this:
+ * - MIDI messages from real device 0 are output to MIDI channels set 0 to 15.
+ * - MIDI messages from real device 1 are output to MIDI channels set 15 to 31.
+ *
+ * 1.2)Now another example with the name "multi:1 0". The driver will forward
+ * MIDI messages as this:
+ * - MIDI messages from real device 1 are output to MIDI channels set 0 to 15.
+ * - MIDI messages from real device 0 are output to MIDI channels set 15 to 31.
+ * So, the device order specified in the setting allows the user to choose the
+ * MIDI channel set associated with this real device at the driver output
+ * according this formula: output_channel = input_channel + device_order * 16.
+ *
+ * 2)Note also that the driver handles single device by putting the device name
+ * in midi.winmidi.device setting.
+ * The user can set the device name "0:Port MIDI SB Live! [CE00]" in the setting.
+ * or use the multi device naming "multi:0" (specifying only device index 0).
+ * Both naming choice allows the driver to handle the same single device.
+ *
  */
 
 #include "fluidsynth_priv.h"
 
 #if WINMIDI_SUPPORT
 
+/* define PRINTF_MSG macro to enable printf message */
+//#define PRINTF_MSG
+
 #include "fluid_midi.h"
 #include "fluid_mdriver.h"
 #include "fluid_settings.h"
 
 #define MIDI_SYSEX_MAX_SIZE     512
 #define MIDI_SYSEX_BUF_COUNT    16
 
-typedef struct
-{
-    fluid_midi_driver_t driver;
-    HMIDIIN hmidiin;
+typedef struct fluid_winmidi_driver_s fluid_winmidi_driver_t;
 
+/* device infos structure for only one midi device */
+typedef struct device_infos_t
+{
+    fluid_winmidi_driver_t *dev; /* driver structure*/
+    unsigned char midi_num;      /* device order number */
+    unsigned char channel_map;   /* MIDI channel mapping from input to output */
+    UINT dev_idx;                /* device index */
+    HMIDIIN hmidiin;             /* device handle */
     /* MIDI HDR for SYSEX buffer */
     MIDIHDR sysExHdrs[MIDI_SYSEX_BUF_COUNT];
+    /* Sysex data buffer */
+    unsigned char sysExBuf[MIDI_SYSEX_BUF_COUNT * MIDI_SYSEX_MAX_SIZE];
+}device_infos;
```

Comment:
The name of this typedef should be `device_infos_t`. Likewise, the name of the struct should not end with `_t`.

Reply:
Ok.